src/persist/persist_utils.hpp

```diff
@@ -178,87 +178,113 @@ class PerThreadCircBuffer{
+            // advance pushed counter.
+            pushed.ui.store(curr_pushed+1, std::memory_order_release);
             return true;
         }
     }
     bool try_pop(const std::function<void(T& x)>& func){
-        int curr_head = head.ui.load(std::memory_order_acquire);
-        if (tail.ui.load(std::memory_order_acquire) == curr_head){
-            // empty.
+        size_t curr_popped = popped.ui.load(std::memory_order_acquire);
+        size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+        assert(curr_popped <= curr_pushed);
+        if (curr_popped == curr_pushed){
+            // empty
             return false;
         }
-        func(payloads[curr_head].ui);
-        head.ui.store(next(curr_head), std::memory_order_release);
-        size--;
+        // consume the to-be-popped entry.
+        func(payloads[curr_popped%cap].ui);
+        // try to pop the next unpopped entry
+        while(!popped.ui.compare_exchange_strong(curr_popped, curr_popped+1, std::memory_order_acq_rel)){
+            size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+            assert(curr_popped <= curr_pushed);
+            if (curr_popped == curr_pushed){
+                // empty
+                return false;
+            }
+        }
+        // successfully popped.
         return true;
     }
     bool try_pop(T& x){
-        int curr_head = head.ui.load(std::memory_order_acquire);
-        if (tail.ui.load(std::memory_order_acquire) == curr_head){
-            // empty.
+        size_t curr_popped = popped.ui.load(std::memory_order_acquire);
+        if (curr_popped == pushed.ui.load(std::memory_order_acquire)){
+            // empty
             return false;
         }
-        x = payloads[curr_head].ui;
-        head.ui.store(next(curr_head), std::memory_order_release);
-        size--;
+        // consume the to-be-popped entry.
+        x = payloads[curr_popped%cap].ui;
+        // try to pop the next unpopped entry
+        while(!popped.ui.compare_exchange_strong(curr_popped, curr_popped+1, std::memory_order_acq_rel)){
+            size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+            assert(curr_popped <= curr_pushed);
+            if (curr_popped == curr_pushed){
+                // empty
+                return false;
+            }
+        }
+        // successfully popped.
         return true;
     }
-    bool full(){
-        return next(tail.ui.load(std::memory_order_acquire)) == head.ui.load(std::memory_order_acquire);
+    void pop_all(const std::function<void(T& x)>& func){
+        while (true){
+            size_t curr_popped = popped.ui.load(std::memory_order_acquire);
+            size_t curr_pushed = pushed.ui.load(std::memory_order_acquire);
+            if (curr_popped == curr_pushed){
+                // empty
+                return;
+            }
+            // consume all entries.
+            size_t i = curr_popped % cap;
```

Comment:
We could assign `i` with `curr_popped` and `end` with `curr_pushed` (perhaps with assert ensuring `curr_pushed` won't exceed `curr_popped` more than `cap`) and write a for loop incrementing `i` to `end`, and consume entry `payloads[i%cap].ui`. I think this would make better readability, though with some negligible overhead of doing mod in every round. 
That said, the current implementation has no issue.

Reply:
You're right; Let's keep it like this for now. Mods every round is exactly what I'm avoiding here.